# About

This directory contains all code that is required to generate results by using the data found in the *Data* directory. A detailed description of each file is found below.

## Shell

### Download.sh
This script downloads all of the required data files to the *Data* direcory and places them in the appropriate subdirectory. This includes 6 sample files, as well as reference genome and gene annotation for *Mycobacterium tuberculosis CDC1551*. 
Note: This script should be run from the *Data* directory.

### Alignment.sh

### HTSeq_count_auto.sh
This script generates a count matrix from the SAM files generated by **Alignment.sh** .and the gene annotation file for *M. tuberculosis CDC1551*. This is done by using the inbuilt function 'htseq-count' from the terminal, and treating it as a black box. The idea is to compare the results yielded by the **DESeq2.R** script when this alternative count matrix is used and when the count matrix generated by the manually written **HTSeq_count.py** is used.
Note: This script should be run from the *Intermediate* directory.

## Python

### HTSeq_count.py
This script generates a count matrix from the SAM files generated by **Alignment.sh** .and the gene annotation file for *M. tuberculosis CDC1551*. This is done through the use of the HTSeq Python package. The count matrix will be used 
Note: This script should be run from the *Intermediate* directory.

### Extract_genes.py
This script extracts a list of gene names from the gene annoatation file for *M. tuberculosis CDC1551*. This list will be used to annotate the count matrix by the **DESeq2.R** script.
Note: This script should be run from the *Intermediate* directory.

## R 

### DESeq2.R
